book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -390,162 +387,183 @@ pub struct NonFinalizedState {
-    - call `let hash = self.push_block_on_chain(block, fork)`
-    - add `fork` to `self.chains`
-    - return `hash`
+#### `pub fn queue(&mut self, new: QueuedBlock)`
 
-5. Else panic, this should be unreachable because `commit_block` is only
-   called when `block` is ready to be committed.
-   
-### `pub(super) fn push_block_on_chain(&mut self, block: QueuedBlock, &mut chain: Chain) -> block::Hash`
+Add a block to the queue of blocks waiting for their requisite context to
+become available.
 
-Try to commit `block` to `chain`. Must succeed, because
-`push_block_on_chain` is only called when `block` is ready to be committed.
+1. extract the `parent_hash`, `new_hash`, and `new_height` from `new.block`
 
-1. push `block` onto `chain`
+2. Add `new` to `self.blocks` using `new_hash` as the key
 
-2. broadcast `result` via `block.rsp_tx`
+3. Add `new_hash` to the set of hashes in
+   `self.by_parent.entry(parent_hash).or_default()`
 
-3. return `block.hash` if `result.is_ok()`
+4. Add `new_hash` to the set of hashes in
+   `self.by_height.entry(new_height).or_default()`
+
+#### `pub fn dequeue_children(&mut self, parent: block::Hash) -> Vec<QueuedBlock>`
+
+Dequeue the set of blocks waiting on `parent`.
+
+1. Remove the set of hashes waiting on `parent` from `self.by_parent`
+
+2. Remove and collect each block in that set of hashes from `self.blocks` as
+  `queued_children`
+
+3. For each `block` in `queued_children` remove the associated `block.hash`
+  from `self.by_height`
+
+4. Return `queued_children`
+
+#### `pub fn prune_by_height(&mut self, finalized_height: block::Height)`
+
+Prune all queued blocks whose height are less than or equal to
+`finalized_height`.
+
+1. Split the `by_height` list at the finalized height, removing all heights
+   that are below `finalized_height`
+
+2. for each hash in the removed values of `by_height`
+    - remove the corresponding block from `self.blocks`
+    - remove the block's hash from the list of blocks waiting on
+      `block.header.previous_block_hash` from `self.by_parent`
 
-4. Else panic, this should be unreachable because `push_block_on_chain` is only
-   called when `block` is ready to be committed.
 
 ### Summary
 
 - `Chain` represents the non-finalized portion of a single chain
 
-- `NonFinalizedState` represents the non-finalized portion of all chains and all
-  unverified blocks that are waiting for context to be available.
+- `NonFinalizedState` represents the non-finalized portion of all chains
+
+- `QueuedBlocks` represents all unverified blocks that are waiting for
+  context to be available.
 
-- `NonFinalizedState::queue` handles queueing and or committing blocks and
-  reorganizing chains (via `commit_block`) but not finalizing them
+- `finalize` must be called manually after calling `commit_block` to prevent
+  chains in `NonFinalizedState` from growing beyond the reorg limit.
 
 - Finalized blocks are returned from `finalize` and must still be committed
   to disk afterwards
```

Comment:
Un-resolved, because the functions in these sections should have explicit names, so they're easier to find:
* https://github.com/ZcashFoundation/zebra/blob/126488e7a3ce2bdfacad6556d99178351fc30423/book/src/dev/rfcs/0005-state-updates.md#committing-non-finalized-blocks
* https://github.com/ZcashFoundation/zebra/blob/126488e7a3ce2bdfacad6556d99178351fc30423/book/src/dev/rfcs/0005-state-updates.md#committing-finalized-blocks

Reply:
Fixed by https://github.com/ZcashFoundation/zebra/pull/1136